doc/source/devel/array.rst

```diff
@@ -0,0 +1,612 @@
+                    enzo_float *zflux, enzo_float dt, 
+                    enzo_float dx, enzo_float dy, enzo_float dz,
+                    int mx, int my, int mz){
+     enzo_float dtdx = dt/dx;
+     enzo_float dtdy = dt/dy;
+     enzo_float dtdz = dt/dz;
+
+     int x_offset = 1;
+     int y_offset = mx;
+     int z_offset = my*mx;
+
+     for (int iz=1; iz<mz-1; iz++) {
+       for (int iy=1; iy<my-1; iy++) {
+         for (int ix=1; ix<mx-1; ix++) {
+           int i = (iz*my + iy)*mx + ix;
+           int i_zf = i;
+           int i_yf = (iz*(my-1) + iy) * mx + ix;
+           int i_xf = (iz*my + iy) * (mx-1) + ix;
+
+           out[i] = (u[i] 
+                     - dtdx * (xflux[i_xf] - xflux[i_xf - x_offset])
+                     - dtdy * (yflux[i_yf] - yflux[i_yf - y_offset])
+                     - dtdz * (zflux[i_zf] - zflux[i_zf - z_offset]));
+         }
+       }
+     }
+   }
+
+
+
+Direction Generalized Functions
+-------------------------------
+
+This example illustrates how subarrays allows functions using
+``CelloArray`` to be written so that they are generalized with respect
+to Cartesian direction. Due to the simplicity of the example, code
+with conventional pointer operations is comparable to the code using
+arrays (however arrays make more complex examples more understandable)
+
+In the van Leer + Constrained Transport scheme, we need to update
+update the cell-centered B-field component along a given direction by
+averaging the same components of the B-field stored at cell
+interfaces. We track Bx at the x-faces, By at the y-faces and Bz at
+the z-faces.
+
+This code assumes a mesh with shape ``(mz, my, mx)``. Suppose we have:
+
+  * An array of cell-centered B-field values (along a given component ) ``bc``
+
+  * An array of interface B-field values (for the same component) ``bi``.
+    This array includes values of cell faces on the exterior of the mesh (e.g.
+    for values centered along the x-axis the shape would be ``(mz,my,mx+1)``).
+
+  * The direction of the component of the B-field is passed in with ``dim``.
+    The values 0,1 & 2 map to x, y, and z
+
+.. code-block:: c++
+
+   typedef double enzo_float;
+   typedef CelloArray<enzo_float,3> EFlt3DArray;
+
+   void calc_center_bfield(EFlt3DArray &bc, EFlt3DArray &bi, int dim){
+     EFlt3DArray bi_l = bi;
+
+     // The following is a repeating pattern that gets factored out into 
+     // a helper function
+     EFlt3DArrau bi_r;
+     if (dim == 0) {
+       bi_r = bi.subarray(CSlice(0,NULL), CSlice(0,NULL), CSlice(1,NULL));
+     } else if (dim == 1) {
+       bi_r = bi.subarray(CSlice(0,NULL), CSlice(1,NULL), CSlice(0,NULL));
+     } else {
+       bi_r = bi.subarray(CSlice(0,NULL), CSlice(1,NULL), CSlice(0,NULL));
+     }
+
+     for (int iz=0; iz<bc.shape(0); iz++) {
+       for (int iy=0; iy<bc.shape(1); iy++) {
+         for(int ix=0; ix<bc.shape(2); ix++) {
+           bc(iz,iy,ix) = 0.5 * (bi_l(iz,iy,ix) + bi_r(iz,iy,ix));
```

Comment:
How about adding a `dim(int dim)` method to `CelloArray`s to be used for loops.
In my mind it is potentially confusing that the dim argument that is going in (and used elsewhere) maps 0,1,2 to x,y,z) but in the loop we'd need to use a map of 0,1,2 to z,y,x.
So there could be a shape that related to the memory layout and a dim that maps to the physical layout.